File: algorithm/Sort/exchange_quick.cpp
/**
 * @Time Complexity:	O(nlog(n))
 * @Space Complexity:	O(nlog(n))
 */

#include <iostream>
#include <stdio.h>

using namespace std;

class Solution {
public:
	static int partition(int unsorted[], int low, int high) {
		int pivot = unsorted[low];
		while (low < high) {
			if (unsorted[high] > pivot)
				high--;
			unsorted[low] = unsorted[high];

			if (unsorted[low] < pivot)
				low++;
			unsorted[high] = unsorted[low];
		}
		unsorted[low] = pivot;
		return low;
	}
	static void quickSort(int unsorted[], int low, int high) {
		int loc = 0;
		if (low < high) {
			loc = partition(unsorted, low, high);
			quickSort(unsorted, low, loc);
			quickSort(unsorted, loc + 1, high);
		}

	}
};

int main() {
	int nums[] = {1, 4, 5, 6, 2, 7, 9, 3};
	Solution sol;
	sol.quickSort(nums, 0, 7);
	for (int i = 0; i < sizeof(nums) / sizeof(nums[0]); i++)
		cout << nums[i] << " ";
	return 0;
}

File: algorithm/Sort/ExchangeQuick.java
package Sort;

/**
 * @Data Structures:
 * @Algorithms: two pointer, partition
 * @Time Complexity:   O(nlogn)
 * @Space Complexity:  O(1)
 */
public class ExchangeQuick {
    private static int partition(int unsorted[], int lo, int hi) {
        int pivot = unsorted[lo];
        while (lo < hi) {
            if (unsorted[hi] > pivot)
                hi--;
            unsorted[lo] = unsorted[hi];

            if (unsorted[lo] < pivot)
                lo++;
            unsorted[hi] = unsorted[lo];
        }
        unsorted[lo] = pivot;
        return lo;
    }

    public static void quickSort(int[] nums, int lo, int hi) {
        if (lo < hi) {
            int loc = partition(nums, lo, hi);
            quickSort(nums, lo, loc);
            quickSort(nums, loc + 1, hi);
        }
    }

    public static void main(String[] args) {
        int[] nums = {5, 2, 3, 1, 6, 7};
        quickSort(nums, 0, nums.length - 1);
        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + ",");
        }
    }
}
